ecs-cli/integ/cmd_up/up_test.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/stretchr/testify/assert"
+	"testing"
+	"time"
+)
+
+const (
+	// timeoutForWaitingOnActiveInstancesInS is how long we are willing wait
+	// for the container instances to become ACTIVE in the cluster.
+	timeoutForWaitingOnActiveInstancesInS = 300
+
+	// sleepDurationInBetweenRetriesInS is how long we sleep in between retrying requests that fail.
+	sleepDurationInBetweenRetriesInS       = 30
+)
+
+// TestClusterCreation runs the 'ecs-cli up -c <clusterName> --capability-iam --force' command.
+// If there is no CloudFormation stack created, then the test fails.
+func TestClusterCreation(t *testing.T) {
+	// Given
+	cfnClient, ecsClient, clusterName := setup(t)
+	cmd := integ.GetCommand([]string{"up", "-c", clusterName, "--capability-iam", "--force"})
+
+	// When
+	stdout, err := cmd.Output()
+	assert.NoError(t, err, fmt.Sprintf("Error running %v\nStdout: %s", cmd.Args, string(stdout)))
+
+	// Then
+	assertHasCFNStack(t, cfnClient, clusterName)
+	assertHasActiveContainerInstances(t, ecsClient, clusterName, 1) // by default we only bring 1 t2.micro instance
+
+	// Cleanup the created resources
+	after(cfnClient, ecsClient, clusterName)
+}
+
+// setup initializes all the clients needed by the test.
+func setup(t *testing.T) (cfnClient *cloudformation.CloudFormation, ecsClient *ecs.ECS, clusterName string) {
+	sess, err := session.NewSession()
+	// Fail the test immediately if we won't be able to evaluate it
+	assert.NoError(t, err, "failed to create new session")
+
+	conf := aws.NewConfig()
+	cfnClient = cloudformation.New(sess, conf)
+	ecsClient = ecs.New(sess, conf)
+	clusterName = integ.SuggestedResourceName()
+	return
+}
+
+// assertHasCFNStack validates that the CFN stack was created successfully
+func assertHasCFNStack(t *testing.T, client *cloudformation.CloudFormation, clusterName string) {
+	resp, err := client.DescribeStacks(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(stackName(clusterName)),
+	})
+	assert.NoError(t, err, "unexpected CloudFormation error during DescribeStacks")
+	assert.NotNil(t, resp.Stacks)
+	assert.Len(t, resp.Stacks, 1)
+	assert.Equal(t, *resp.Stacks[0].StackName, stackName(clusterName))
+}
+
+// assertHasActiveContainerInstances validates that the containers in the cluster are all eventually ACTIVE
+func assertHasActiveContainerInstances(t *testing.T, client *ecs.ECS, clusterName string, clusterSize int) {
+	maxNumRetries := timeoutForWaitingOnActiveInstancesInS / sleepDurationInBetweenRetriesInS
+	for retryCount := 0; retryCount < maxNumRetries; retryCount++ {
+		cluster, err := client.ListContainerInstances(&ecs.ListContainerInstancesInput{
+			Cluster: aws.String(clusterName),
+		})
+		if err != nil || len(cluster.ContainerInstanceArns) != clusterSize {
+			t.Log("No available container instances in the cluster, retry...")
```

Comment:
to make this reusable in cases where we expect more than 1 instance, can we change the above check to `len(...) < clusterSize` and use the msg "Expected number of instances in cluster not found, retrying..." ?

Reply:
`!=` or `<` would serve the same purpose here I think. Anyway, instead of lumping these two checks in the same if-statement I've split them up so that we can have more granular error logs.